src/gc.c

```diff
@@ -2111,6 +2115,39 @@ JL_DLLEXPORT void *jl_gc_managed_realloc(void *d, size_t sz, size_t oldsz,
     return b;
 }
 
+JL_DLLEXPORT jl_value_t *jl_gc_realloc_string(jl_value_t *s, size_t sz)
+{
+    size_t len = jl_string_len(s);
+    if (sz <= len) return s;
+    size_t strsz = len + sizeof(size_t) + 1;
+    if (strsz <= GC_MAX_SZCLASS) {
+        // pool allocated; can't be grown in place so allocate a new object.
+        jl_value_t *snew = jl_alloc_string(sz);
+        memcpy(jl_string_data(snew), jl_string_data(s), len);
+        return snew;
+    }
+    size_t newsz = sz + sizeof(size_t) + 1;
+    size_t offs = offsetof(bigval_t, header);
+    size_t allocsz = LLT_ALIGN(newsz + offs, JL_CACHE_BYTE_ALIGNMENT);
+    if (allocsz < sz)  // overflow in adding offs, size was "negative"
+        jl_throw(jl_memory_exception);
+    bigval_t *hdr = bigval_header(jl_astaggedvalue(s));
+    gc_big_object_unlink(hdr);
+    // TODO: this is not safe since it frees the old pointer. ideally we'd like
+    // the old pointer to be left alone if we can't grow in place.
+    // for now it's up to the caller to make sure there are no references to the
+    // old pointer.
+    bigval_t *newbig =
```

Comment:
Hmmm, We may also need some special handling here. If a GC is triggered in this realloc the object will not be in any list. This means that the unlinking and linking code in the marking will be very sad and it'll also miss the sweep. Missing the sweep is fortunately not too bad since we'll just skip it the next time and there's no pointer to be scanned in this object anyway.

Making linking/unlinking thread safe should be able to fix this with some tweaking too (it queues the requrest) Best stop gap solution for now is probably having a version of `realloc` that does not do gc and call `maybe_collect` before unlinking.